### PR TITLE
Language and mathematical standardizations

### PR DESCRIPTION
**Pull request by jmason42**
This addresses #10, #16, and #19.  I also slightly redesigned the propensity recalculation logic, as it was recomputing some propensities multiple times in the same step, and it was also recomputing propensities even if only the product abundances changed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,62 +1,81 @@
...
-    [-2, 0, 0, 1],
-    [1, 1, -1, 0]])
+# Each column is a reaction and each row is a molecular species (or other
+# entity). The first reaction here means that the first and second elements
+# combine to create the third, while the fourth is unaffected.
+stoichiometry = np.array([
+    [-1, -2, +1],
+    [-1,  0, +1],
+    [+1,  0, -1],
+    [ 0, +1,  0]
```
Comment (prismofeverything) [inline README.md]: I gotta say, this transposed form is unpleasant to me, though if it is some kind of standard I understand. But having each reaction split out among every row upsets some organizational principle deep in my soul.

File: arrow/arrow.py [hunk 2]
```diff
@@ -9,84 +9,90 @@ def choose(n, k):
...
 
     return np.array(terms).prod()
 
-
-def step(reactions, rates, state, forms, propensities=[], update_reactions=()):
+def step(stoichiometry, rates, state, forms, propensities=[], update_reactions=()):
     if len(update_reactions):
         for update in update_reactions:
-            reaction = reactions[update]
+            reaction_stoichoiometry = stoichiometry[:, update]
```
Comment (prismofeverything) [inline arrow/arrow.py]: My only issue with this whole `reaction_stoichiometry` vs `stoichiometry` vs `reaction` business is that if we are using `reaction_stoichiometry` then the other `stoichiometry` should be comparably qualified, perhaps `system_stoichiometry`, to make it conceptually consistent. However, that gets cumbersome everywhere. 

My other issue is that `reaction_stoichiometry` is long : ) I know `reaction` as a shorthand for this concept is perhaps not exactly precise, but I think it makes sense in context here. Also, not so into the word `stoichiometry` standing in for both all reactions and a single reaction. Any other alternatives here?

File: arrow/arrow.py [hunk 3]
```diff
@@ -9,84 +9,90 @@ def choose(n, k):
...
+            stoichiometry,
             rates,
             state,
             forms,
             propensities,
             update_reactions)
 
-        time += dt
-        if not dt or time > duration:
+        t += dt
```
Comment (prismofeverything) [inline arrow/arrow.py]: Not sure about this substitution of `time` for `t`.... in general I try to avoid single letter variable names or abbreviations in general (though `dt` is established enough to evade this tendency).

File: arrow/arrow.py (hunk 2 above)
Comment (jmason42) [reply arrow/arrow.py]: Yeah, the language is ugly.  After this pass I'm inclined to go back and distinguish the `stoichiometric_matrix` (the full system) versus a `stoichiometry` (which is reaction specific).  The former is verbose but explicit.

File: arrow/arrow.py (hunk 3 above)
Comment (jmason42) [reply arrow/arrow.py]: Right, I was a bit annoyed by this too.  I was trying to keep the internal and external meaning of `time` consistent and also liked the parallelism of `t` and `dt`.  Will think on this.

File: README.md (hunk 1 above)
Comment (jmason42) [reply README.md]: There are a few different things colliding here.  In biochemical dynamics literature we write `dc/dt = Sv` where `Sv` is the matrix-vector product between the stoichiometric matrix and a vector of reaction rates; in this form, `S` has to have reactions along the columns and metabolites along the rows.  This is how `S` is defined in the FBA literature, too.

However, there's an issue - we don't just use the stoichiometry to decide what a reaction does, we also use it to decide the rate of reaction (`v`), since that is dependent upon reactant abundances.  So stoichiometry is actually serving two different purposes; first, to calculate rates (our propensities), and second, to decide what the update actually is, given that an event occurs (this is the true meaning of stoichiometry).

On top of all that, we're naturally doing things on a per-reaction basis, since Gillespie's SSA is designed such that only one reaction occurs at each step.  Thus we don't really get much out of the standard dynamic form.

So, in short: this *is* the standard form of a stoichiometric matrix, even if it's inconvenient and inefficient for this particular problem.  I think we should drop the matrix representation anyway (#8), since in practice these systems are so sparse (#9), at which point this becomes a non-issue.  I can prioritize that change.

File: arrow/arrow.py (hunk 2 above)
Comment (prismofeverything) [reply arrow/arrow.py]: Hmmm that seems like a good middle ground.

File: arrow/arrow.py (hunk 3 above)
Comment (prismofeverything) [reply arrow/arrow.py]: Yeah I can see maybe `time` is a misnomer, as it does not represent the entirety of time, but rather a time point. I can also see using `now`.

File: README.md (hunk 1 above)
Comment (prismofeverything) [reply README.md]: Awesome, thanks for the explanation. So this is a classic mismatch between theoretical forms and what makes sense in a computer. I would say we can present the interface in the dynamic way that everyone expects, and do the computation with the transpose which makes more sense for the computer to do. Retrieving the reaction stoichiometry with a simple index into the whole matrix seems too natural not to use, but yeah the sparse representation would make all of this moot anyway.

Comment (jmason42) [reply README.md]: What's really odd is that I've tried switching to a Fortran memory ordering, and if anything the evaluation time goes up.  It's hard to say because the variability in each evaluation is on the order of maybe 20% of the average run time.  I'm going to leave it alone for now.

**Comment by jmason42** [comment]
These changes should address your concerns.  I went with `time_current` in place of `t` and `time_to_next` in place of `dt`, both of which are clumsy but at least clearer and more consistent.

**Comment by prismofeverything** [comment]
Awesome, thanks for this! I'll roll it into a new release.
